Add a Transport::TCP class for tcp-based communication.

andy31415 opened the pull request:
#### Problem
CHIP requires the ability to communicate over TCP (for key negotiations, for faster data transfers when supported (e.g. OTA)). 

 #### Summary of Changes
Implements a TCP-based transport, with equivalent functionality as the current Transport::UDP.

fixes #2664

andy31415 commented on src/transport/PeerAddress.h:
```diff
@@ -96,11 +97,15 @@ class PeerAddress
     {
-        return (mTransportType == other.mTransportType) && (mIPAddress == other.mIPAddress) && (mPort == other.mPort);
+        return (mTransportType == other.mTransportType) && (mIPAddress == other.mIPAddress) && (mPort == other.mPort) &&
+            ((mInterface == INET_NULL_INTERFACEID) || (other.mInterface == INET_NULL_INTERFACEID) ||
```
I think we had missed the inteface compare before.

But it seems to make sense to me to have null be treated as 'any' for a compare. If we do not like that, we need more thought into how TCP determines 'an active connection already exists'.

Likely peer communication still needs some thought - we can likely compare IP address however port can be anything since it is random on connection establishment on one side.

andy31415 replied:
Ok, updated. 
Our unit tests currently use null interface id, so some bugs may creep up.

BroderickCarlin commented on src/transport/PeerAddress.h:
```diff
@@ -48,7 +48,7 @@ enum class Type
     kUndefined,
     kUdp,
     kBle,
-    // More constants to be added later, such as TCP
```
Should we maintain this comment or is TCP the last constant we plan to add?

andy31415 replied:
hmmm ... I had removed this in some intermediate versions. Should be fixed now.

andy31415 replied:
I am not aware of anything else we could add as chip spec. I believe a comment would be 'add more stuff here if needed' however that seems redundant as any enum would be 'add more if needed'.

bzbarsky-apple commented on src/transport/TCP.h:
```diff
@@ -0,0 +1,235 @@
+/**
+ *    @file
+ *      This file defines the CHIP Connection object that maintains TCP connections.
+ *      It binds to any avaiable local addr and port and begins listening.
```
"available"

bzbarsky-apple commented on src/transport/TCP.cpp:
```diff
@@ -0,0 +1,585 @@
+
+/**
+ *    @file
+ *      This file implements the CHIP Connection object that maintains a TCP connection.
```
It maintains multiple connections, no?

bzbarsky-apple commented on src/transport/TCP.h:
```diff
@@ -0,0 +1,235 @@
+        mActiveConnectionsSize(bufferSize), mPendingPackets(packetBuffers), mPendingPacketsSize(packetsBuffersSize)
+    {
+        PendingPacket emptyPending{
+            .peerAddress  = PeerAddress::Uninitialized(),
```
Designated initializers in C++ are non-standard until C++20, no?  I'm a little surprised this compiled on all our platforms, actually.

andy31415 replied:
If it compiles everywhere, should I bother to update it as this is quite readable?

bzbarsky-apple commented on src/transport/TCP.h:
```diff
@@ -0,0 +1,235 @@
+    bool HasActiveConnections() const;
+
+    /**
+     * Close all active connetions
```
"connections"

bzbarsky-apple commented on src/transport/TCP.h:
```diff
@@ -0,0 +1,235 @@
+     * @param addr - what peer to connect to
+     * @param msg - what buffer to send once a connection has been established.
+     */
+    CHIP_ERROR SendAfterConnect(const PeerAddress & addr, System::PacketBuffer * msg);
```
Should probably document that this function takes ownership of the buffer.

bzbarsky-apple commented on src/transport/TCP.h:
```diff
@@ -0,0 +1,235 @@
+     * @param peerAddress the peer the data is coming from
+     * @param buffer the actual data
+     */
+    CHIP_ERROR ProcessReceivedBuffer(Inet::TCPEndPoint * endPoint, const PeerAddress & peerAddress, System::PacketBuffer * buffer);
```
Again, should document that the function takes ownership of the buffer.

bzbarsky-apple commented on src/transport/TCP.h:
```diff
@@ -0,0 +1,235 @@
+     * @param buffer the buffer containing the message
+     * @param messageSize how much of the head contains the actual message
+     */
+    CHIP_ERROR ProcessSingleMessageFromBufferHead(const PeerAddress & peerAddress, System::PacketBuffer * buffer,
```
And here should document what the ownership semantics for buffer are, because I think that would make it clearer whether the caller and the implementation agree on them.

andy31415 replied:
I updated the code to correctly handle ownership. Also found the case where no handler was set was not clearning buffers.

bzbarsky-apple commented on src/transport/TCP.h:
```diff
@@ -0,0 +1,235 @@
+    // Called when a connection has been completed
+    static void OnConnectionComplete(Inet::TCPEndPoint * endPoint, INET_ERROR err);
+
+    // Called when a connection has been closed
```
Might be worth documenting how this differs from `OnPeerClosed`...

andy31415 replied:
I updated this mentioning this originates from TCPEndPoint. 
That class has two callbacks. I am actually unsure why - peerclosed is on read 0 (peer closed) however I am unclear how this one is called (OnClose and OnAbort actually null-out the callbacks with the comment 'application called close explicitly')

bzbarsky-apple commented on src/transport/Tuple.h:
```diff
@@ -129,6 +131,27 @@ class Tuple : public Base
+    /**
+     * Recursive disconnect implementation iterating through transport members.
+     *
+     * @tparam N the index of the underlying transport to send disconnect to
```
"@param", right?

andy31415 replied:
tparam is correct for template parameters.

bzbarsky-apple commented on src/transport/tests/TestTCP.cpp:
```diff
@@ -0,0 +1,216 @@
+const char PAYLOAD[]        = "Hello!";
+int ReceiveHandlerCallCount = 0;
+
+void MessageReceiveHandler(MessageHeader & header, const Transport::PeerAddress & source, System::PacketBuffer * msgBuf,
```
We really need to document the ownership API contracts for this stuff.  I would _assume_ that this should take ownership of the incoming buffer, freeing it before returning, like `SecureSessionMgrBase::HandleDataReceived`does.  But the API on `Transport::Base` does not actually document the expected behavior of the function that gets passed to `SetMessageReceiveHandler`, so maybe it's the secure session manager that's doing this wrong?

andy31415 replied:
At least for TCP, buffers may contain several messages so it is not correct for handlers to free the data. I will review the usages to see if this is maintained within UDP and such.

bzbarsky-apple commented on src/transport/tests/TestTCP.cpp:
```diff
@@ -0,0 +1,216 @@
+    err = tcp.SendMessage(header, Transport::PeerAddress::TCP(addr), buffer);
+    if (err == System::MapErrorPOSIX(EADDRNOTAVAIL))
+    {
+        // TODO: the underlying system does not support IPV6. This early return should
```
Please file an issue tracking this.

andy31415 replied:
This is a copy and paste from TestUDP ... at that time we had errors in CI, which should already have tracking issues.
I will update the wording.

andy31415 replied:
Added an issue after all, since I believe the original issue for CI is already addressed in a curent PR which likely will not touch TestTCP/UDP.

Opened issue is #2698 and is now referenced in the TODO.

bzbarsky-apple commented on src/transport/TCP.cpp:
```diff
@@ -0,0 +1,585 @@
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Inet, "Connection complete encountered an error: %s", ErrorStr(err));
+        endPoint->Free();
```
And in general, maybe the manipulation of the combination of `mActiveConnections` and `mUsedEndPointCount` should move into a helper object that ensures that it maintains whatever invariants we want here.

andy31415 replied:
Decremented.

the endpoint count seems indeed error prone. I would update it as a secondary pass though as the change is already larger than I initially expected.Async connection establishment really complicates things.

andy31415 commented on src/transport/TCP.h:
```diff
@@ -0,0 +1,235 @@
+    static void OnPeerClosed(Inet::TCPEndPoint * endPoint);
+
+    // Called when a connection is received on the listening port
+    static void OnConnectionRecevied(Inet::TCPEndPoint * listenEndPoint, Inet::TCPEndPoint * endPoint,
```
I cannot spot the typo here... what did I miss?

bzbarsky-apple replied:
"recevied" vs "received".  The order of 'v' and 'i' is wrong.

andy31415 replied:
heh ... I had pasted both and could not see it. Small letters! 
Thanks .. updating.

andy31415 replied:
Fixed. Funnily enough searching for recevied finds stuff in some 3rd party repos (bluez and nxp).

andy31415 commented on src/transport/TCP.cpp:
```diff
@@ -0,0 +1,585 @@
+    SuccessOrExit(err);
+
+    buffer->ConsumeHead(headerSize);
+    HandleMessageReceived(header, peerAddress, buffer);
```
:( This needs review then, thank you for spotting.

I don't think we can free the message as is since TCP may contain multiple messages. I will update comments and add an extra retain.

andy31415 replied:
I changed the code to AddRef, assuming the handle will do a free.

andy31415 commented:
@bzbarsky-apple I have addressed all comments. Will double check on freeing paths ... I wish we had some 'smart pointer' like constructs instead of raw points. Tracking buffer usage manually seems error prone to me. Maybe topic for the future.

andy31415 commented:
> @bzbarsky-apple I have addressed all comments. Will double check on freeing paths ... I wish we had some 'smart pointer' like constructs instead of raw pointers. Tracking buffer usage manually seems error prone to me. Maybe topic for the future.

SecureSession does indeed Free the passed in buffer, so right now all our paths seem to behave correctly (after my updates to the test).

andy31415 commented on src/transport/TCP.cpp:
```diff
@@ -0,0 +1,585 @@
+            VerifyOrExit(buffer->DataLength() >= messageSize, err = CHIP_ERROR_INTERNAL);
+
+            err = ProcessSingleMessageFromBufferHead(peerAddress, buffer, messageSize);
+            buffer->ConsumeHead(messageSize);
```
I updated the code so that process ensures buffer is not freed (tcp needs to handle its own freeing since a buffer may contain multiple messages)

andy31415 commented on src/transport/TCP.cpp:
```diff
@@ -0,0 +1,585 @@
+
+        if ((inetErr == CHIP_NO_ERROR) && (err == CHIP_NO_ERROR))
+        {
+            err = endPoint->Send(tcp->mPendingPackets[i].packetBuffer);
```
Added a free. Send frees implicityly and I added a free in case we are not even attempting to free.

andy31415 commented on src/transport/TCP.cpp:
```diff
@@ -0,0 +1,585 @@
+    {
+        for (size_t i = 0; i < tcp->mActiveConnectionsSize; i++)
+        {
+            if (tcp->mActiveConnections[i] == nullptr)
```
Our logic should ensure this would never happen, however I added a check with log in case we do have logic errors. The count tracking is a bit on the complex side, especially in unexpected error conditions (happy path is unit tested already)

woody-apple commented:
@saurabhst ?

bzbarsky-apple commented:
> I wish we had some 'smart pointer' like constructs instead of raw pointers. 

Yes, indeed.  I filed https://github.com/project-chip/connectedhomeip/issues/2707

bzbarsky-apple commented on src/transport/TCP.h:
```diff
@@ -0,0 +1,244 @@
+                                                  uint16_t messageSize);
+
+    // Callback handler for TCPEndPoint. TCP message receive handler.
+    static void OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuffer * buffer);
```
This should probably document ownership semantics too...

andy31415 replied:
Should be handled by TCPEndPoint. That documents it as:

```
     *  A data reception event handler must acknowledge data processed using
     *  the \c AckReceive method. The \c Free method on the data buffer must
     *  also be invoked unless the \c PutBackReceivedData is used instead
```

I think we should not try to duplicate docs lest they become out of sync.

bzbarsky-apple replied:
OK, sounds good!  Could add a pointer to the TCPEndPoint documentation from here.

andy31415 commented on src/transport/TCP.cpp:
```diff
@@ -0,0 +1,617 @@
+            VerifyOrExit(buffer->DataLength() >= messageSize, err = CHIP_ERROR_INTERNAL);
+
+            err = ProcessSingleMessageFromBufferHead(peerAddress, buffer, messageSize);
+            buffer->ConsumeHead(messageSize);
```
I thought to keep it because I alrady skipped over the size. If processsinglemessage fails on a buffer, there seems to be reasonable expectation that it will fail in the future as well. This also guarantees that we empty buffers in time.

I.e. I think this is correct and it would be more work to not consume on error.

bzbarsky-apple replied:
OK.  That seems fine, but worth a comment explaining why we are doing what we are doing.

andy31415 commented:
Created #2708 for better active count usage.
